File: Algorithms/Search/Binary Search/BinarySearch.js
const BinarySearchRecursion = (
  arr,
  target,
  left = 0,
  right = arr.length - 1
) => {
  if (left > right) return "Not Found";
  let middle = Math.floor((left + right) / 2);
  if (arr[middle] < target) {
    return BinarySearchRecursion(arr, target, middle + 1, right);
  } else if (arr[middle] > target) {
    return BinarySearchRecursion(arr, target, left, middle - 1);
  } else {
    return middle;
  }
};
const binarySearchLoops = (arr, target, left = 0, right = arr.length - 1) => {
  while (left <= right) {
    let middle = Math.floor((left + right) / 2);
    if (target > arr[middle]) {
      left = middle + 1;
    } else if (target < arr[middle]) {
      right = middle - 1;
    } else if (target == arr[middle]) {
      return middle;
    }
  }
  return -1;
};
// Time comlexity: O(log n)
console.log(binarySearchLoops([-1, 0, 3, 5, 9, 12], 2));

File: Algorithms/Search/Binary Search/BinarySearch.py
import math;
def binary_search(arr, target, left = 0 , right = None ):
    if right == None:
        right = len(arr) - 1
    middle = math.floor((left + right) / 2) 
    if(left > right):   
        return 'Not found'
    if ( arr[middle] > target ):    
        return binary_search(arr, target, left, middle - 1 )
    if ( arr[middle] < target):
        return binary_search(arr, target, middle + 1, right )

    return middle;

print(binary_search([1,2,3,4,5,6,7],3))